Pagination

jackdclark opened the pull request:
Closes: #41, closes #42, and closes #29 

~Provides a `fetchMore` function that will send the same query with the provided options merged into the original options, meaning the consumer doesn't have to manually maintain the state.~ Removed this as the same can be achieved with `refetch` now that it accepts `options`.

`useQuery` accepts an `updateData` function as part of the `options` object that enables control of how the previous & new results are merged, this function can be overridden if necessary when calling `refetch` as well.

### TODO
- [x] Agree on implementation
- [x] Test coverage
- [x] Update documentation

### Example usage

```jsx
// use options.updateResult to append the new page of posts to our current list of posts
const updateData = (prevData, prevData) => ({
  ...data,
  allPosts: [...prevData.allPosts, ...data.allPosts]
});

export default function PostList() {
  // set a default offset of 0 to load the first page
  const [skipCount, setSkipCount] = useState(0);

  const { loading, error, data } = useQuery(
    allPostsQuery,
    { variables: { skip: skipCount, first: 10 } },
    updateData
  );

  if (error) return <div>There was an error!</div>;
  if (loading && !data) return <div>Loading</div>;

  const { allPosts, _allPostsMeta } = data;
  const areMorePosts = allPosts.length < _allPostsMeta.count;

  return (
    <section>
      <ul>
        {allPosts.map(post => (
          <li key={post.id}>
            <a href={post.url}>{post.title}</a>
          </li>
        ))}
      </ul>
      {areMorePosts && (
        <button
          // set the offset to the current number of posts to fetch the next page
          onClick={() => setSkipCount(allPosts.length)}
        >
          Show more
        </button>
      )}
    </section>
  );
}
```

bmullan91 commented on src/useClientRequest.js:
```diff
@@ -57,7 +57,7 @@ function useClientRequest(query, initialOpts = {}) {
 
   // arguments to fetchData override the useClientRequest arguments
-  async function fetchData(newOpts) {
+  async function fetchData(newOpts, updateResult) {
```
What if we made `updateResult` part of the `options` object?

jackdclark replied:
@Joezo suggested the same, will move it 👍

bmullan91 commented on src/useQuery.js:
```diff
@@ -25,11 +23,21 @@ module.exports = function useQuery(query, opts = {}) {
 
   React.useEffect(() => {
-    queryReq();
+    queryReq(null, updateResult);
```
If the `update` function was part of the `options` we wouldn't need to do this

bmullan91 commented on src/useQuery.js:
```diff
@@ -25,11 +23,21 @@ module.exports = function useQuery(query, opts = {}) {
+        },
+        updateResultOverride || updateResult
+      );
+    }
```
This would become:

```js
fetchMore: (fetchMoreOptions) => queryReq(fetchMoreOptions)
```

We don't need to merge the options here - they are already bound to the `queryReq` by the closure in useClientRequest which does the merging/overriding.

jackdclark replied:
In this merge I was also doing a deeper merge on `options.variables` so that only the ones that are changing need to be passed in each time. As discussed, I've updated this to only do the merge on the variables & pass the rest through.

bmullan91 commented on src/useQuery.js:
```diff
@@ -7,13 +7,11 @@ const defaultOpts = {
 };
 
-module.exports = function useQuery(query, opts = {}) {
+module.exports = function useQuery(query, opts = {}, updateResult) {
```
If `update` was part of the options we don't need to update the `useQuery` signature.

bmullan91 commented on src/useQuery.js:
```diff
@@ -30,6 +28,10 @@ module.exports = function useQuery(query, opts = {}) {
+    fetchMore: ({ variables, ...rest } = {}) =>
+      // merge variables so they don't all have to be passed back in,
+      // just the ones that are changing
+      queryReq({ ...rest, variables: { ...allOpts.variables, ...variables } })
```
😍

bmullan91 commented on test/unit/useClientRequest.test.js:
```diff
@@ -212,7 +212,91 @@ describe('useClientRequest', () => {
+
+        expect(fetchData({ variables: { limit: 20 } })).rejects.toThrow(
+          'options.updateResult must be a function'
+        );
```
👏 excellent tests

jackdclark commented:
@bmullan91 as discussed earlier I've removed `fetchMore` in favour of the declarative implementation. If a more imperative style is desired then the same can be achieved by using `refetch` and passing in new options.

I've also added a pagination guide to the README that includes distinct page & infinite loading examples.

If you could take another look when you get a chance please, hopefully we can get this in next week 😀

bmullan91 commented on src/useClientRequest.js:
```diff
@@ -88,7 +88,14 @@ function useClientRequest(query, initialOpts = {}) {
+      if (typeof revisedOpts.updateResult !== 'function') {
+        throw new Error('options.updateResult must be a function');
+      }
+      result.data = revisedOpts.updateResult(state.data, result.data);
```
Since the name is `updateResult`, maybe we should pass the entire result? It might be useful for a developer to change / override the errors in some circumstances.

`useClientRequest` would then look something like:
```js
let result = await client.request(revisedOperation, revisedOpts);

if (revisedOpts.updateResult) {
  if (typeof revisedOpts.updateResult !== 'function') {
    throw new Error('options.updateResult must be a function');
  }
  result = revisedOpts.updateResult(state, result)
}
```

The pagination example `updateResult`:
```js
const updateResult = (prevResult, newResult) => ({
  ...newResult,
  data: {
    ...newResult.data,
    allPosts: [...prevResult.data.allPosts, ...result.data.allPosts]
  }
});

useQuery(MY_QUERY, {
  updateResult
})
```

It makes updating the data a bit more verbose, but the developer could write a wrapper if it was a really common usecase:

```js
const dataUpdater = (fn) => (prevResult, newResult) => ({
  ...newResult,
  data: fn(prevResult.data, newResult.data)
})
```

jackdclark replied:
As discussed, passing the entire result back allows internal properties like `loading` & `error` to be changed, or new ones to be added which could be abused or cause tricky bugs. We're going to rename this to `updateData` as a better description of what it's doing & if we see a use case for changing / overriding errors in the future, we could always introduce an `updateErrors` option.

bmullan91 commented on README.md:
```diff
@@ -309,6 +314,130 @@ The `options` object that can be passed either to `useMutation(mutation, options
+      count
+    }
+  }
+`;
```
We should simplify the example - this is could be:

```js
ALL_POSTS_QUERY = `
  query($limit: Int, $skip: Int) {
    allPosts(limit: $limit, skip: $skip) {
      total,
      items {
        id
        title
      }
    }
  }
`
```

jackdclark replied:
I think the separation between data & meta is good here? There are redundant parts to this example that could be removed & make it more readable though, how about:

```javascript
export const ALL_POSTS_QUERY = `
  query allPosts($first: Int!, $skip: Int!) {
    allPosts(first: $first, skip: $skip) {
      id
      title
      url
    }
    _allPostsMeta {
      count
    }
  }
`;
```

bmullan91 commented on src/useQuery.js:
```diff
@@ -30,6 +28,11 @@ module.exports = function useQuery(query, opts = {}) {
+    refetch: (options = {}) =>
+      queryReq({
+        skipCache: true,
+        updateData: (_, data) => data,
```
Minor: We should add a comment here saying why we're doing this

jackdclark replied:
Agree, I've added an explanation
